code/espurna/sensors/BME680Sensor.h

```diff
@@ -0,0 +1,242 @@
+        // Public
+        // ---------------------------------------------------------------------
+
+        BME680Sensor() {
+            _sensor_id = SENSOR_BME680_ID;
+            _count = 9;
+        }
+
+        // ---------------------------------------------------------------------
+        // Sensor API
+        // ---------------------------------------------------------------------
+
+        void begin() {
+          if (!_dirty) {
+            return;
+          }
+
+          // I2C auto-discover
+          unsigned char addresses[] = {BME680_I2C_ADDR_PRIMARY, BME680_I2C_ADDR_SECONDARY};
+          _address = _begin_i2c(_address, sizeof(addresses), addresses);
+          if (_address == 0) return;
+
+          iaqSensor.begin(_address, Wire);
+
+          #if SENSOR_DEBUG
+            DEBUG_MSG("[BME680] BSEC library version v%u.%u.%u.%u\n",
+              iaqSensor.version.major,
+              iaqSensor.version.minor,
+              iaqSensor.version.major_bugfix,
+              iaqSensor.version.minor_bugfix
+            );
+          #endif
+
+          if (!_isSensorOk()) {
+            _error = SENSOR_ERROR_OTHER;
+            return;
+          }
+
+          iaqSensor.setConfig(bsec_config_iaq);
+          iaqSensor.updateSubscription(sensorList, 12, BSEC_SAMPLE_RATE_LP);
+
+          if (!_isSensorOk()) {
+            _error = SENSOR_ERROR_OTHER;
+            return;
+          }
+
+          _ready = true;
+          _dirty = false;
+        }
+
+        // Descriptive name of the sensor
+        String description() {
+            char buffer[21];
+            snprintf(buffer, sizeof(buffer), "BME680 @ I2C (0x%02X)", _address);
+            return String(buffer);
+        }
+
+        // Type for slot # index
+        unsigned char type(unsigned char index) {
+            if (index == 0) return MAGNITUDE_TEMPERATURE;
+            if (index == 1) return MAGNITUDE_HUMIDITY;
+            if (index == 2) return MAGNITUDE_PRESSURE;
+            if (index == 3) return MAGNITUDE_RESISTANCE;
+            if (index == 4) return MAGNITUDE_IAQ_ACCURACY;
+            if (index == 5) return MAGNITUDE_IAQ_STATIC;
+            if (index == 6) return MAGNITUDE_IAQ;
+            if (index == 7) return MAGNITUDE_CO2E;
+            if (index == 8) return MAGNITUDE_BVOC;
+
+            return MAGNITUDE_NONE;
+        }
+
+        // The maximum allowed time between two `bsec_sensor_control` calls depends on
+        // configuration profile `bsec_config_iaq` below. With these settings, the
+        // maximum allowed time is 3s.
+        void tick() {
+            static unsigned long last_millis = 0;
+
+            if (millis() - last_millis < BME680_READ_INTERVAL) return;
+
+            last_millis = millis();
+
+            _error = SENSOR_ERROR_OK;
+
+            if (!_isSensorOk()) {
+              _error = SENSOR_ERROR_OTHER;
+              return;
+            }
+
+            if (iaqSensor.run()) {
```

Comment:
btw I noticed run() already tries to be called only once per reading through the use of millis() + overflow counter (see getTimeMs()) and nextCall timestamp.

Meaning, there might be no need for additional BME680_READ_INTERVAL?
e.g. if `nextCall - getTimeMs()` is the time in millis to sleep, if run() succeeded we can create a new timer to sleep for the specified time. see `Ticker::once_ms`
https://github.com/esp8266/Arduino/tree/master/libraries/Ticker/examples

Reply:
@mcspr I'm a bit lost :) are you suggesting something like this?

```cpp
        void tick() {
            ticker.attach(2000, _read)
        }

        void _read() (
          _error = SENSOR_ERROR_OK;

          if (!_isSensorOk()) {
            _error = SENSOR_ERROR_OTHER;
            return;
          }

          if (iaqSensor.run()) {
            _rawTemperature = iaqSensor.rawTemperature;
            _rawHumidity = iaqSensor.rawHumidity;
            _temperature = iaqSensor.temperature;
            _humidity = iaqSensor.humidity;
            _pressure = iaqSensor.pressure / 100;
            _gasResistance = iaqSensor.gasResistance;
            _iaqAccuracy = iaqSensor.iaqAccuracy;
            _iaq = iaqSensor.iaq;
            _iaqStatic = iaqSensor.staticIaq;
            _co2Equivalent = iaqSensor.co2Equivalent;
            _breathVocEquivalent = iaqSensor.breathVocEquivalent;
          }
        )
```

Reply:
Actually, I think we can simply call `iaqSensor.run()` on `tick()` without any other guards.

Reply:
Or, that. The idea with ticker was about removing run() calculation of diff each loop() call, and instead only do it after successful run(). But, not using _read as ticker callback, but simply adding a boolean flag to do it tick(). Otherwise, b/c Ticker callback is called inside of system context, bsec lib won't like that.

Just a hunch, might be not *that* bad. Quick perf debug:
```cpp
auto before = ESP.getCycleCount();
run();
auto after = ESP.getCycleCount();
DEBUG_MSG_P(PSTR("[perf] run() took %u µs\n"), clockCyclesToMicroseconds(after - before));
```